sdks/python/apache_beam/runners/portability/universal_local_runner.py

```diff
@@ -0,0 +1,398 @@
+        self._messages.append(message)
+    threading.Thread(target=read_messages).start()
+
+    for state_response in self._handle.GetStateStream(
+        beam_job_api_pb2.GetJobStateRequest(job_id=self._job_id)):
+      self._state = self._runner_api_state_to_pipeline_state(
+          state_response.state)
+      if state_response.state in TERMINAL_STATES:
+        break
+    if self._state != runner.PipelineState.DONE:
+      raise RuntimeError(
+          "Pipeline %s failed in state %s." % (self._job_id, self._state))
+
+
+class BeamJob(threading.Thread):
+  """This class handles running and managing a single pipeline.
+  """
+  def __init__(self, job_id, pipeline_options, pipeline_proto,
+               use_grpc=True, sdk_harness_factory=None):
+    super(BeamJob, self).__init__()
+    self._job_id = job_id
+    self._pipeline_options = pipeline_options
+    self._pipeline_proto = pipeline_proto
+    self._use_grpc = use_grpc
+    self._sdk_harness_factory = sdk_harness_factory
+    self._log_queue = queue.Queue()
+    self._state_change_callbacks = [
+        lambda new_state: self._log_queue.put(
+            beam_job_api_pb2.JobMessagesResponse(
+                state_response=
+                beam_job_api_pb2.GetJobStateResponse(state=new_state)))
+    ]
+    self._state = None
+    self.state = beam_job_api_pb2.JobState.STARTING
+    self.daemon = True
+
+  def add_state_change_callback(self, f):
+    self._state_change_callbacks.append(f)
+
+  @property
+  def log_queue(self):
+    return self._log_queue
+
+  @property
+  def state(self):
+    return self._state
+
+  @state.setter
+  def state(self, new_state):
+    for state_change_callback in self._state_change_callbacks:
+      state_change_callback(new_state)
+    self._state = new_state
+
+  def run(self):
+    with JobLogHandler(self._log_queue):
+      try:
+        fn_api_runner.FnApiRunner(
+            use_grpc=self._use_grpc,
+            sdk_harness_factory=self._sdk_harness_factory
+        ).run_via_runner_api(self._pipeline_proto)
+        self.state = beam_job_api_pb2.JobState.DONE
+      except:  # pylint: disable=bare-except
+        logging.exception("Error running pipeline.")
+        traceback.print_exc()
+        self.state = beam_job_api_pb2.JobState.FAILED
+
+  def cancel(self):
+    if self.state not in TERMINAL_STATES:
+      self.state = beam_job_api_pb2.JobState.CANCELLING
+      # TODO(robertwb): Actually cancel...
+      self.state = beam_job_api_pb2.JobState.CANCELLED
+
+
+class JobServicer(beam_job_api_pb2.JobServiceServicer):
+  """Servicer for the Beam Job API.
+
+  Manages one or more pipelines, possibly concurrently.
+  """
+  def __init__(
+      self, worker_command_line=None, use_grpc=True):
+    self._worker_command_line = worker_command_line
+    self._use_grpc = use_grpc or bool(worker_command_line)
+    self._jobs = {}
+
+  def start_grpc(self, port=0):
+    self._server = grpc.server(futures.ThreadPoolExecutor(max_workers=3))
```

Comment:
<!--new_thread; commit:77e51af3ee2af39264c366a2602957e7e9bb20c7; resolved:0-->
Curious, why 3 is a good number here?

Reply:
No reason, allows for some parallelism, but doesn't spawn a huge number of threads.